extension/json/src/utils/json_utils.cpp

```diff
@@ -642,60 +642,97 @@ uint32_t jsonArraySize(const JsonWrapper& wrapper) {
     return yyjson_arr_size(yyjson_doc_get_root(wrapper.ptr));
 }
 
-static bool jsonContains(yyjson_val* haystack, yyjson_val* needle);
-
-static bool jsonContainsArr(yyjson_val* haystack, yyjson_val* needle) {
-    uint32_t haystackIdx = 0, haystackMax = 0;
-    yyjson_val* haystackChild = nullptr;
-    yyjson_arr_foreach(haystack, haystackIdx, haystackMax, haystackChild) {
-        if (jsonContains(haystackChild, needle)) {
-            return true;
-        }
+// Helper function to check scalar values for strict equality
+bool jsonContainsScalar(yyjson_val* haystack, yyjson_val* needle) {
+    // Only return true if types and values are equal
+    if (yyjson_get_type(haystack) != yyjson_get_type(needle)) {
+        return false;
     }
-    return false;
+    return yyjson_equals(haystack, needle);
 }
 
-static bool jsonContainsArrArr(yyjson_val* haystack, yyjson_val* needle) {
-    uint32_t needleIdx = 0, needleMax = 0, haystackIdx = 0, haystackMax = 0;
-    yyjson_val *needleChild = nullptr, *haystackChild = nullptr;
+// Function to check if a needle array is contained within a haystack array
+bool jsonContainsArrArr(yyjson_val* haystack, yyjson_val* needle) {
+    uint32_t needleIdx = 0, needleMax = 0;
+    yyjson_val* needleChild = nullptr;
     yyjson_arr_foreach(needle, needleIdx, needleMax, needleChild) {
         bool found = false;
+        uint32_t haystackIdx = 0, haystackMax = 0;
+        yyjson_val* haystackChild = nullptr;
         yyjson_arr_foreach(haystack, haystackIdx, haystackMax, haystackChild) {
             if (jsonContains(haystackChild, needleChild)) {
                 found = true;
+                break; // Stop searching once a match is found
             }
         }
         if (!found) {
-            return jsonContainsArr(haystack, needle);
+            return false; // Return false if any needle element is missing
         }
     }
     return true;
 }
 
-static bool jsonContainsObj(yyjson_val* haystack, yyjson_val* needle) {
+// Function to check if any element in a haystack array matches the needle
```

Comment:
I suggest moving the comments into the  header file

Reply:
Good advice, moved the description of each function to the json_util.h